### PR TITLE
Bugfix/28924 fix large module upgrade from zip

### LINKED ISSUE #28924: BO - An unclear error is displayed when upgrading a too large module using the zip
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Thanks, @PrestaShop/qa-functional team for your help in investigating this issue :heart: 
To test a release of any module, I need to upgrade the module using a zip
But, in the 80x branch, an error is displayed when we use a **too large** module (because we launched the `npm install`) 
To create a wishlist zip from Github, we follow these steps:

1.  git clone git@github.com:PrestaShop/blockwishlist.git
2. cd blockwishlist
3. `git prc origin [id_pr] dev `or you can use those commands `git fetch origin pull/IdPR/head:pr-IdPR && git checkout pr-IdPR && git reset --hard && git clean -fd && git rebase origin/dev && composer install`
4. **npm install**
5. Create a zip

Now, in the BO > Modules > Module manager, when we try to upgrade to this new zip by uploading it => an error is displayed
In some cases, we have this error
image
In other cases, we have this error
image

https://user-images.githubusercontent.com/16067358/177360536-f4f64b12-34bc-46a0-b962-b5d8e6633ade.mp4

I checked my configuration

- phpinfo file 

image

- BO > Advanced Parmeters > Administration page 

image

### Expected behavior
In the PS1786 > we have this alert 
image

### Steps to reproduce

1. Install 8.0.x
2. create a zip of any PR of the module [blockwishlist](https://github.com/PrestaShop/blockwishlist) following these steps:
1.  git clone git@github.com:PrestaShop/blockwishlist.git
2. cd blockwishlist
3. git prc origin [id_pr] dev
4. **npm install**
5. Create a zip
6. Go to BO > Modules > Module manager > Upload the new zip
7. See error

### PrestaShop version(s) where the bug happened

8.0.x

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by MeKeyCool**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Error message was wrong when tried to upload too big module zip file. (See below for more details)
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28924 .
| Related PRs       | Las working commit was [ea89268](https://github.com/PrestaShop/PrestaShop/commit/ea89268) but at this time, check was done by a front end default value of module size limit. Did back-end check ever worked ? Then front-end has to manage bad error message as well.   
| How to test?      | Cf. #28924 
| Possible impacts? | Please check "Hooks" panel (debug toolbar) for all important pages.

> :metal: special thanks to @FabienPapet for the hand about symfony profiler / datacollector architecture and to @jolelievre for last "serialize" analysis xD.

Without this fix, my development environment throw a `Symfony\\Component\\ErrorHandler\\Error\\OutOfMemoryError` exception due to profiler / data collector tooling that induced some weird json response.

My environment :
```ini
; Maximum amount of memory a script may consume (128MB)
; http://php.net/memory-limit
memory_limit = 256M
```
After some analysis, we can see that symfony request object is given as hook and module arguments.
Cf. [HookManager](https://github.com/PrestaShop/PrestaShop/blob/8.0.x/src/Adapter/HookManager.php#L71) and [HookDispatcher](https://github.com/PrestaShop/PrestaShop/blob/8.0.x/src/Adapter/Hook/HookDispatcher.php#L238)

## Detailed description

> :point_up:  In fact this bug as probably always been there.

It has been revealed when module size check has stopped being done by front-end.
Cf. https://github.com/PrestaShop/PrestaShop/commit/8a6765299a1c586c774f130e43ce024ca8cc6dad : DropZone library configuration changed when this line has been removed : `maxFilesize: 50, // can't be greater than 50Mb because it's an addons limitation`

Then the back end returned a valid error message but during [`index.php` `kernel::terminate()` operation](https://github.com/PrestaShop/PrestaShop/blob/8.0.x/admin-dev/index.php#L83), an error occurred during [Profiler](https://symfony.com/doc/4.4/profiler.html) data collected [serialization](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpKernel/Profiler/FileProfilerStorage.php#L160) and added some weird json stack trace to API response.
> Then the response looked like : 
> ```json
> {
>     API expected error message
> } {
>     kernel::terminate() error stack trace (OutOfMemory)
> }
> ```

Two problems were found there : 
1. Front end didn't manage properly invalid json response from API.
2. The custom [HookDataCollector](https://github.com/PrestaShop/PrestaShop/blob/8.0.x/src/PrestaShopBundle/DataCollector/HookDataCollector.php) added all called and not called hooks and modules with their parameters. The main element that as been identified as memory waste is the "request" argument that is already given in debug bar and given by default to all modules and hooks (cf. [HookManager](https://github.com/PrestaShop/PrestaShop/blob/8.0.x/src/Adapter/HookManager.php#L71) and [HookDispatcher]https://github.com/PrestaShop/PrestaShop/blob/8.0.x/src/Adapter/Hook/HookDispatcher.php#L238))

> :warning: Even if this change fix the problem and guaranty some memory usage optimization for debug, it is not normal that an Exception could break API response. I recommend some further explorations about error handling management with API.

### CODE REVIEW

File: src/PrestaShopBundle/DataCollector/HookDataCollector.php [hunk 1]
```diff
@@ -140,6 +140,23 @@ private function stringifyHookArguments(array &$hooksList)
...
             }
         }
+        // foreach ($hooksList as $hookList) {
+        //     foreach ($hookList as $hook) {
+        //         $hook['args'] = $this->cloneVar($hook['args']);
+
+        //         foreach ($hook['modules'] as $modulesByType) {
+        //             foreach ($modulesByType as $type => $module) {
+        //                 if (empty($module)) {
+        //                     unset($modulesByType[$type]);
```
Comment (matks) [inline src/PrestaShopBundle/DataCollector/HookDataCollector.php]: Commented code should not be submitted in PR 😉

Comment (MeKeyCool) [reply src/PrestaShopBundle/DataCollector/HookDataCollector.php]: Sorry, until @boubkerbribri helped me, Iwas fighting with CI to understand what is wrong ^^'
I think this fix hasn't finished to torture me yet :sob: ...

File: src/PrestaShopBundle/DataCollector/HookDataCollector.php [hunk 2]
```diff
@@ -125,15 +125,26 @@ private function stringifyHookArguments(array &$hooksList)
...
                             unset($modulesByType[$type]);
+                            continue;
                         }
 
                         if (array_key_exists('args', $module)) {
+                            // Symfony Request object is quite heavy and is already given by debug tooling. Don't clone it again.
+                            if (isset($module['args']['request'])) {
+                                $module['args']['request'] = 'Module use client request as argument. For more details, cf. "Request/Response panel".';
+                            }
+
```
Comment (atomiix) [inline src/PrestaShopBundle/DataCollector/HookDataCollector.php:147]: It doesn't look like this is related to the original bug, is it? 🤔

Comment (jolelievre) [reply src/PrestaShopBundle/DataCollector/HookDataCollector.php:147]: @atomiix absolutely not 😅

It was a really weird bug, actually, the message for file size was indeed sent in the response, but after that the kernel called its `terminate` method, which goes through this HookCollector (which is used by the debug bar) and then it went out of memory.

When that happens there's an automatic error handler that appended another JSON message stating that memory went out of bound, but since it was appended after an initial JSON object the final result was unparsable.

```json
{
  error: 'File too big'
} {
  error: 'No more memory'
}
```

@MeKeyCool I think you need to describe all this in the PR's description because a reviewer can't guess it on first sight. And even for future dev who would track back the reason of this modification it's important to explain the context thoroughly in the description.

As to why it didn't happen previously, my guess is that the large file uses a big part of the memory in itself which means not much is left at the end of the process. So it doesn't mean this will never happen again, it just means the risks have been reduced.

Comment (jolelievre) [reply src/PrestaShopBundle/DataCollector/HookDataCollector.php:147]: So the compromise we chose was to not display the Request in the Hokk parameters panel, the request is passed to almost all hooks, meaning the object is cloned several times in a process which is overkill since the debug bar already has a panel allowing to check the full Request with more details.

File: themes/package.json [hunk 3]
```diff
@@ -33,4 +33,4 @@
         "webpack": "^5.37.1",
         "webpack-cli": "^4.7.0"
     }
-}
\ No newline at end of file
+}
```
Comment (jolelievre) [inline themes/package.json]: You should revert this change

**Review by jolelievre — DISMISSED** [review]
Thanks @MeKeyCool

**Comment by MeKeyCool** [comment]
@atomiix and @jolelievre is this issue detailed enough now ?

File: src/PrestaShopBundle/DataCollector/HookDataCollector.php (hunk 2 above)
Comment (atomiix) [reply src/PrestaShopBundle/DataCollector/HookDataCollector.php:147]: So if I understand well, this bug only happens in debug mode, right?

Comment (MeKeyCool) [reply src/PrestaShopBundle/DataCollector/HookDataCollector.php:147]: I think so :man_shrugging:

**Review by jolelievre — APPROVED** [review]
Thanks @MeKeyCool

**Review by atomiix — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hello @PrestaShop/prestashop-maintainers,

As mentioned in the issue https://github.com/PrestaShop/PrestaShop/issues/28924#issuecomment-1294872355, the bug is not reproduced anymore.
Can we close this PR?

Thank you!

**Comment by kpodemski** [comment]
Hello @khouloudbelguith 

The provided video doesn't answer whether there is a bug or not. There's no error while uploading a "too big file"? There should be, right? I feel like the uploaded package was valid and wasn't too big unless I don't understand this PR and the problem in general.

**Comment by khouloudbelguith** [comment]
Hello @kpodemski,

Yes, I checked again, I didn't reproduce the issue using the branch 8.0.x.
I used the blockwishlist module clone from this repository https://github.com/PrestaShop/blockwishlist 

When uploading the zip => I have a successful alert 
image

But when using the PS1.7.8.7, we cannot upload the zip 
image

I checked my configuration

- phpinfo file 

image

- BO > Advanced Parmeters > Administration page 

image

Ping @PrestaShop/product-team what do you think?
We have a different behavior between PS1787 & PS8.0.x.
What is the expected behavior in this case?
PS: the original issue is not reproduced anymore with 8.0.x branch.

Thanks!

**Comment by kpodemski** [comment]
@khouloudbelguith the point of this fix was to fix the behavior when you upload a file that is too big, maybe you should get your limits much lower values and then try to upload a file?

for instance, limit 1M, file 2M

**Review by MhiriFaten — COMMENTED** [review]
Hello @MeKeyCool ,

I checked your PR and I reproduce the issue using the branch **8.0.x**
I used the blockwishlist module clone from this repository https://github.com/PrestaShop/blockwishlist and a `blockwishlist` file ZIP `587,1kB` 

[Untitled_ Jan 16, 2023 9_57 AM.webm](https://user-images.githubusercontent.com/111756615/212637536-84234244-b3b8-4f4a-ba5f-08b8e59ecb24.webm)

Can you check it please ?

**Comment by kpodemski** [comment]
I tested a PR, and there's always a warning:

`<b>Warning</b>:  POST Content-Length of 18130335 bytes exceeds the limit of 8388608 bytes in <b>Unknown</b> on line <b>0</b><br />
`

So it doesn't really work. I'm closing it as there's no chance for the original author to continue work.
